### DISCUSSION TITLE
ssh-agent seems broken

### DISCUSSION THREAD

**Discussion by jgonggrijp** [blinksh/blink#1844] [Support] [answered]
I have trouble using agent forwarding with iPadOS 16.6 and Blink 16.1.12.800 connecting over SSH to a Mac mini with macOS 15.6 (remote `home`). 

Steps after a fresh startup of Blink, three attempts in total. I suspect the final output might hold a clue to what is wrong.

```
blink> ssh-add -l
# no output
blink> ssh -A home
Connected to ...
Last login: ...
julian@home ~ % ssh-add -l
4096 SHA256:XUildWaQ6N9tk/x6n3cguZEtPTxVtAhI6YNeOZcFsUI id_rsa (RSA)
4096 SHA256:XUildWaQ6N9tk/x6n3cguZEtPTxVtAhI6YNeOZcFsUI id_rsa (RSA)
julian@home ~ % cd a/dir/with/a/git/repo
julian@home ~ % git fetch
Enter passphrase for key '/Users/julian/.ssh/id_rsa': 

julian@home ~ % exit
blink> ssh-add
Key id_rsa - added to agent.
blink> ssh-add -l
4096 SHA256:XUildWaQ6N9tk/x6n3cguZEtPTxVtAhI6YNeOZcFsUI id_rsa (RSA)
blink> ssh -A home
Connected to ...
Last login: ...
julian@home ~ % ssh-add -l
4096 SHA256:XUildWaQ6N9tk/x6n3cguZEtPTxVtAhI6YNeOZcFsUI id_rsa (RSA)
4096 SHA256:XUildWaQ6N9tk/x6n3cguZEtPTxVtAhI6YNeOZcFsUI id_rsa (RSA)
4096 SHA256:XUildWaQ6N9tk/x6n3cguZEtPTxVtAhI6YNeOZcFsUI id_rsa (RSA)
julian@home ~ % cd a/dir/with/a/git/repo
julian@home ~ % git fetch
Enter passphrase for key '/Users/julian/.ssh/id_rsa': 

julian@home ~ % exit
blink> ssh-agent
Failed loading ssh_agent_main from MAIN, cause = dlsym(RTLD_MAIN_ONLY, ssh_agent_main): symbol not found
ssh-agent: command not found
```

**Comment by carloscabanero** [comment]
Hey @jgonggrijp, sorry for not jumping on this sooner. We recently changed the way that an existing key in the agent will get handled if it is the same key as used to sign in. Would you mind to update and let us know if this is now fixed from your side?

**Comment by jgonggrijp** [comment]
Blink version 16.2.1.826, identical results.

**Reply by carloscabanero** [reply, marked as answer]
Ok, so let me start with how to do what you are want to, and then I will explain why your suggestion does not work (although it maybe should).

Blink runs one Host Agent and one "global" or Default Agent (the one you control with ssh-add). The way to do what you want is to add the key to the Host Agent instead of the Default Agent. To do so, go to Hosts -> <host> and in the SSH Agent section, select Agent Forwarding and the key you want to pass. You can select a "Confirm" constraint for the key, so Blink will ask you every time it gets requested by the remote (more secure), or just do "Always" for unconstrained access. This has multiple advantages, the key will only be forwarded to that host, you don't have to do ssh-add all the time and ssh -A is done on your behalf too.

Now, on why. Blink uses multiple agents as it is generally considered more secure. Multiple agents allows us to identify which target is asking for a key, and that is how we are able to "Constraint" keys and even show a prompt: "The host X is trying to use the key Y, do you allow it?". This is similar to what [ssh-ident](https://github.com/ccontavalli/ssh-ident) does. Each Host Agent is linked to the Default Agent (the one you used with ssh-add). This allows to forward specific keys for a host, but then still have a "global agent", for example with a Git key that can be forwarded to multiple places. The ssh-agent protocol is really barebones, and all these "tricks" are improvements that usually people build on-top but we built them as part of our agent.

The problem you found is that you are adding to the Default agent **the same key** that you used to sign in (the per Host Agent). **So the key is in the Host Agent (constrained only for logging in) and in the Default Agent, and that is why you see it multiple times.** When you are trying to use it, the Agent will deny using the key, as the key in the Host Agent is used first and it can only be used for logging in. When you add it as I described, then that constraint is removed. But you do not need to add it to the default agent.

So we broke the convention (or the expectation) of how things are supposed to work in this scenario vs the barebones and global ssh-agent. I will create an Issue as fixing this should not be hard. But truth is, from a product perspective, Blink does not need the "default" agent anymore for most cases. We should consider to hide it or to make this behavior more obvious somehow.

Anyway, hope this was not a lot, let me know if you have any other questions about it and how it goes.

**Reply by carloscabanero** [reply]
Fixed the regular behavior here: https://github.com/blinksh/blink/issues/1849

Should be out on next TF v17.0.0

**Reply by jgonggrijp** [reply]
> The way to do what you want is to add the key to the Host Agent instead of the Default Agent. To do so, go to Hosts -> and in the SSH Agent section, select Agent Forwarding and the key you want to pass. You can select a "Confirm" constraint for the key, so Blink will ask you every time it gets requested by the remote (more secure), or just do "Always" for unconstrained access. This has multiple advantages, the key will only be forwarded to that host, you don't have to do ssh-add all the time and ssh -A is done on your behalf too.

While I didn't mention it, I tried this, too, and it didn't work. But I will double-check later today whether this is still the case.

**Reply by carloscabanero** [reply]
I tried this yesterday while doing #1849 and it was working without issue. Try it again when you get the chance and let me know.

**Reply by jgonggrijp** [reply]
I tried, and I hereby confirm that the problem is still there. My steps:

1. Go to settings, then hosts, then the `home` host.
2. "Agent Forwarding" was already set to "Confirm". I changed it to "Always" just to rule out interactions between different bugs.
3. "Forward keys" was already set to the key I wanted to forward.
4. Click "Save" in the top right corner of the host config page.
5. Click the back button in the top left until I can go back no further (in case this is needed to really truly actually store the settings).
6. Click outside of the settings pane to close it.
7. Purge Blink from memory and reopen.
8. Check that "Agent Forwarding" is still set to "Always" and "Forward Keys" is still set to the key I want to forward.
9. `ssh home`
10. `ssh-add -l` lists my key twice (I don't mind it it does, by the way, but I mention it for completeness).
11. I am prompted for the password of my SSH key on the remote host when I try to use SSH, just like in the example output I posted before.

**Reply by carloscabanero** [reply]
Hmmmmm... ok. I'm unable to replicate this at the moment. So this may sound stupid, but let's first discard an issue with your other remote configuration. The way we usually test this is to connect to the server and the server to itself using the same key through the agent. 

So add the key for "home" as part of the agent forward (make sure it is the only key so we reduce variables). Then connect from blink to home, and then from home to localhost - obviously make sure you do not have other keys at "home" to connect to itself. 

Does that go through?

**Reply by jgonggrijp** [reply]
Yes it does! And for the first time, I actually saw this confirmation, because I switched Agent Forwarding back to "Confirm" in the meanwhile.

`home` connected to itself, I needed to confirm the remote host key, and I needed to call `exit` twice before I was back in the Blink shell. I double-checked the `.ssh/authorized_keys` on `home` and it doesn't contain any of the public keys on `home`.

**Reply by carloscabanero** [reply]
Ok! So we know the agent is working and it is forwarding keys. Now we need to figure out why the other remote is not accepting the key. To do so, I will need you to send me the verbose output from "home" to the remote: "ssh -vvv <second_remote>". Usually the sequence we are looking for here is: 1) ssh will indicate that it is talking to our agent and it will request the public keys; 2) then it will see if any of those public keys are installed in the remote; 3) and then if any of them match it will be instructed by the other remote to use it for authentication.

Some things that can go wrong here: the obvious is that the public key is not installed in the remote, then maybe the public key type is not accepted by the remote (ie, some platforms are very picky), etc...

**Reply by jgonggrijp** [reply]
I was doing this, but I had an external screen connected in mirror mode. When the output appeared, Blink broke out of mirror mode into this mode where the external screen has a separate Blink window (scale, probably). The window I was doing `ssh -vvv other-remote` in is on the built-in screen of my ipad, while the forwarding confirmation is on my external screen. Alas, the mouse cannot move onto that screen and it is not a touchscreen. The window on the iPad does not respond to ctrl-c, cmd-., space or return, so I'm stuck.

I realize this is a separate issue, but do you know of a way to get out of this deadlock so I can retry testing without an external screen connected?

**Reply by jgonggrijp** [reply]
I should mention I tried switching off the screen, but unfortunately that does not move the confirmation dialog to the ipad screen.

**Reply by carloscabanero** [reply]
Hmmm... good catch that Blink would break out of mirror mode when the pop up showed. We will check that.

Closing Blink -> Disconnecting the Screen -> Reopening Blink should do the trick. There should be no other window where the Agent can create the pop up in this case and Blink should not have any state stored.

**Reply by jgonggrijp** [reply]
Alright, it was working all along. Sorry for the noise.

When I just ran into this deadlock with the confirmation dialog on the wrong screen, I realized that agent forwarding was working; otherwise I would not see the dialog. I then wondered what was the difference with my previous attempts. That difference was easy; I was previously trying `git fetch` in one of the local repos on the `home` remote. You asked me to do `ssh -vvv`, so I needed a different experiment.

The new experiment was `ssh -T -vvv git@github.com`. It produces the confirmation dialog and agent forwarding works.

The local repo I was previously testing `git fetch` in, had a remote clone on codeberg.org, where I had not added my Blink key yet.

Again, sorry for the noise!

**Reply by carloscabanero** [reply]
No worries at all. Glad we figured this out, we discovered two real issues, and you gave us a few ideas for future QA testing!  Noise is good :)
